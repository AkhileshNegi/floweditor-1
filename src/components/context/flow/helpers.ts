import { snakify } from '~/utils';

import { FlowResult, FlowResultMap } from './FlowContext';

export const getFlowResult = (
    results: FlowResultMap,
    resultName: string,
    nodeUUID: string
): FlowResult => {
    const key = snakify(resultName);
    const result = results[key] || { key, name: resultName, nodeUUIDs: [] };
    result.nodeUUIDs = result.nodeUUIDs.filter((uuid: string) => uuid !== nodeUUID);
    result.nodeUUIDs.push(nodeUUID);
    return result;
};
