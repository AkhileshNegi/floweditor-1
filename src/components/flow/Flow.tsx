import { react as bindCallbacks } from 'auto-bind';
import * as React from 'react';
import { EditorConsumer, EditorState } from '~/components/context/editor/EditorContext';
import { ConnectionEvent, FlowConsumer, FlowState } from '~/components/context/flow/FlowContext';
import * as styles from '~/components/flow/Flow.scss';
import ConnectedNode from '~/components/flow/node/Node';
import { getDraggedFrom } from '~/components/helpers';
import ConnectedNodeEditor from '~/components/nodeeditor/NodeEditor';
import Simulator from '~/components/simulator/Simulator';
import Sticky from '~/components/sticky/Sticky';
import { ConfigProviderContext } from '~/config';
import { fakePropType } from '~/config/ConfigProvider';
import { getActivity } from '~/external';
import { FlowDefinition } from '~/flowTypes';
import ActivityManager from '~/services/ActivityManager';
import Plumber from '~/services/Plumber';
import { DragSelection } from '~/store/editor';
import { RenderNodeMap } from '~/store/flowContext';
import { getCollisions } from '~/store/helpers';
import { contextTypes } from '~/testUtils';
import {
    createUUID,
    isRealValue,
    NODE_PADDING,
    renderIf,
    snapToGrid,
    timeEnd,
    timeStart
} from '~/utils';
import Debug from '~/utils/debug';

export interface FlowProps {
    definition: FlowDefinition;
    flowState?: FlowState;
    editorState?: EditorState;
}

export interface Translations {
    [uuid: string]: any;
}

export const REPAINT_TIMEOUT = 500;
export const GHOST_POSITION_INITIAL = { left: -1000, top: -1000 };

export const nodeSpecId = 'node';
export const nodesContainerSpecId = 'node-container';
export const ghostNodeSpecId = 'ghost-node';
export const dragSelectSpecId = 'drag-select';

export const isDraggingBack = (event: ConnectionEvent) => {
    return event.suspendedElementId === event.targetId && event.source !== null;
};

export const getDragStyle = (drag: DragSelection) => {
    const left = Math.min(drag.startX, drag.currentX);
    const top = Math.min(drag.startY, drag.currentY);
    const width = Math.max(drag.startX, drag.currentX) - left;
    const height = Math.max(drag.startY, drag.currentY) - top;
    return {
        left,
        top,
        width,
        height
    };
};

export class Flow extends React.Component<FlowProps> {
    private Activity: ActivityManager;
    private Plumber: Plumber;
    private containerOffset = { left: 0, top: 0 };
    private ele: HTMLDivElement;
    private nodeContainerUUID: string;

    // Refs
    private ghost: any;

    public static contextTypes = {
        endpoints: fakePropType,
        debug: fakePropType
    };

    constructor(props: FlowProps, context: ConfigProviderContext) {
        super(props, context);

        this.nodeContainerUUID = createUUID();

        this.Activity = new ActivityManager(props.flowState.definition.uuid, getActivity);

        this.Plumber = new Plumber();

        /* istanbul ignore next */
        if (context.debug) {
            window.fe = new Debug(props.editorState, this.props.editorState.debug);
        }

        bindCallbacks(this, {
            include: [/Ref$/, /^on/, /^is/]
        });

        timeStart('RenderAndPlumb');
    }

    private onRef(ref: HTMLDivElement): HTMLDivElement {
        return (this.ele = ref);
    }

    private ghostRef(ref: any): any {
        return (this.ghost = ref);
    }

    public componentDidMount(): void {
        this.Plumber.bind('connection', (event: ConnectionEvent) =>
            this.props.flowState.mutator.updateConnection(event.sourceId, event.targetId)
        );

        this.Plumber.bind('beforeDrag', (event: ConnectionEvent) =>
            this.beforeConnectionDrag(event)
        );

        this.Plumber.bind('connectionDrag', (event: ConnectionEvent) => {
            this.props.flowState.mutator.updateConnectionDrag(event);
        });

        this.Plumber.bind('connectionDragStop', (event: ConnectionEvent) =>
            this.onConnectorDrop(event)
        );

        this.Plumber.bind(
            'beforeStartDetach',
            (event: ConnectionEvent) => !this.props.editorState.translating
        );

        this.Plumber.bind('beforeDetach', (event: ConnectionEvent) => true);

        this.Plumber.bind('beforeDrop', (event: ConnectionEvent) =>
            this.onBeforeConnectorDrop(event)
        );

        // If we don't have any nodes, create our first one
        this.props.flowState.mutator.ensureStartNode();

        let offset = { left: 0, top: 0 };

        /* istanbul ignore next */
        if (this.ele) {
            offset = this.ele.getBoundingClientRect();
        }
        this.containerOffset = { left: offset.left, top: offset.top + window.scrollY };

        timeEnd('RenderAndPlumb');

        // deals with safari load rendering throwing
        // off the jsplumb offsets
        window.setTimeout(() => this.Plumber.repaint(), REPAINT_TIMEOUT);
    }

    public componentWillUnmount(): void {
        this.Plumber.reset();
    }

    /**
     * Called right before a connector is dropped onto a new node
     */
    private onBeforeConnectorDrop(event: ConnectionEvent): boolean {
        this.props.flowState.mutator.closeNodeEditor(true);

        if (event.sourceId.split(':')[0] === event.targetId) {
            console.error(event.targetId + ' cannot point to itself');
            return false;
        }

        return true;
    }

    // private onShowDefinition(definition: FlowDefinition): void {
    //     TODO: make this work, it's cool!
    //     this.Plumber.reset();
    //     this.setState({ viewDefinition: definition }, () => { this.Plumber.repaint() });
    // }

    /**
     * Called the moment a connector is done dragging, whether it is dropped on an
     * existing node or on to empty space.
     */
    private onConnectorDrop(event: ConnectionEvent): boolean {
        const { ghostNode } = this.props.flowState;
        // Don't show the node editor if we a dragging back to where we were
        if (isRealValue(ghostNode) && !isDraggingBack(event)) {
            // Wire up the drag from to our ghost node
            this.Plumber.recalculate(ghostNode.node.uuid);

            const dragPoint = getDraggedFrom(ghostNode);
            this.Plumber.connect(
                dragPoint.nodeUUID + ':' + dragPoint.exitUUID,
                ghostNode.node.uuid
            );

            // Save our position for later
            const { left, top } = snapToGrid(this.ghost.ele.offsetLeft, this.ghost.ele.offsetTop);

            ghostNode.ui.position = { left, top };
            let originalAction = null;
            if (ghostNode.node.actions && ghostNode.node.actions.length === 1) {
                originalAction = ghostNode.node.actions[0];
            }

            // Bring up the node editor
            this.props.flowState.mutator.openNodeEditor({
                originalNode: ghostNode,
                originalAction
            });
        }

        // To-do: mock this out
        /* istanbul ignore next */
        $(document).off('mousemove');

        return true;
    }

    private beforeConnectionDrag(event: ConnectionEvent): boolean {
        return !this.props.editorState.translating;
    }

    private getNodes(nodes: RenderNodeMap): JSX.Element[] {
        return Object.keys(nodes).map(uuid => {
            const renderNode = nodes[uuid];
            return (
                <ConnectedNode
                    key={uuid}
                    data-spec={nodeSpecId}
                    renderNode={renderNode}
                    Activity={this.Activity}
                    plumberRepaintForDuration={this.Plumber.repaintForDuration}
                    plumberDraggable={this.Plumber.draggable}
                    plumberMakeTarget={this.Plumber.makeTarget}
                    plumberRemove={this.Plumber.remove}
                    plumberRecalculate={this.Plumber.recalculate}
                    plumberMakeSource={this.Plumber.makeSource}
                    plumberConnectExit={this.Plumber.connectExit}
                    plumberSetDragSelection={this.Plumber.setDragSelection}
                    plumberClearDragSelection={this.Plumber.clearDragSelection}
                    plumberRemoveFromDragSelection={this.Plumber.removeFromDragSelection}
                />
            );
        });
    }

    private getStickies(): JSX.Element[] {
        const stickyMap = this.props.flowState.definition._ui.stickies || {};
        return Object.keys(stickyMap).map(uuid => {
            return (
                <Sticky
                    key={uuid}
                    uuid={uuid}
                    sticky={stickyMap[uuid]}
                    plumberClearDragSelection={this.Plumber.clearDragSelection}
                    plumberDraggable={this.Plumber.draggable}
                    plumberRemove={this.Plumber.remove}
                />
            );
        });
    }

    private getDragNode(): JSX.Element {
        return isRealValue(this.props.flowState.ghostNode) ? (
            <ConnectedNode
                ref={this.ghostRef}
                key={this.props.flowState.ghostNode.node.uuid}
                data-spec={ghostNodeSpecId}
                ghost={true}
                renderNode={this.props.flowState.ghostNode}
                Activity={this.Activity}
                plumberRepaintForDuration={this.Plumber.repaintForDuration}
                plumberDraggable={this.Plumber.draggable}
                plumberMakeTarget={this.Plumber.makeTarget}
                plumberRemove={this.Plumber.remove}
                plumberRecalculate={this.Plumber.recalculate}
                plumberMakeSource={this.Plumber.makeSource}
                plumberConnectExit={this.Plumber.connectExit}
                plumberClearDragSelection={this.Plumber.clearDragSelection}
                plumberSetDragSelection={this.Plumber.setDragSelection}
                plumberRemoveFromDragSelection={this.Plumber.removeFromDragSelection}
            />
        ) : null;
    }

    private getSimulator(): JSX.Element {
        return renderIf(this.context.endpoints && this.context.endpoints.simulateStart)(
            <Simulator Activity={this.Activity} plumberDraggable={this.Plumber.draggable} />
        );
    }

    private getNodeEditor(): JSX.Element {
        return renderIf(this.props.flowState.nodeEditorOpen)(
            <ConnectedNodeEditor
                plumberConnectExit={this.Plumber.connectExit}
                plumberRepaintForDuration={this.Plumber.repaintForDuration}
            />
        );
    }

    private isClickOnCanvas(event: React.MouseEvent<HTMLDivElement>): boolean {
        // TODO: not sure the TS-safe way to access id here
        return (event.target as any).id === this.nodeContainerUUID;
    }

    private onDoubleClick(event: React.MouseEvent<HTMLDivElement>): void {
        if (this.isClickOnCanvas(event)) {
            const { left, top } = snapToGrid(
                event.clientX - this.containerOffset.left - 100 + NODE_PADDING,
                event.clientY - this.containerOffset.top - NODE_PADDING * 2
            );

            this.props.flowState.mutator.updateSticky(createUUID(), {
                position: { left, top },
                title: 'New Note',
                body: '...'
            });
        }
    }

    public onMouseDown(event: React.MouseEvent<HTMLDivElement>): void {
        if (this.isClickOnCanvas(event)) {
            this.props.editorState.mutator.mergeEditorState({
                dragSelection: {
                    startX: event.pageX - this.containerOffset.left,
                    startY: event.pageY - this.containerOffset.top,
                    currentX: event.pageX - this.containerOffset.left,
                    currentY: event.pageY - this.containerOffset.top,
                    selected: null
                }
            });
        }
    }

    public onMouseMove(event: React.MouseEvent<HTMLDivElement>): void {
        if (this.props.editorState.dragSelection && this.props.editorState.dragSelection.startX) {
            const drag = this.props.editorState.dragSelection;
            const left = Math.min(drag.startX, drag.currentX);
            const top = Math.min(drag.startY, drag.currentY);
            const right = Math.max(drag.startX, drag.currentX);
            const bottom = Math.max(drag.startY, drag.currentY);

            this.props.editorState.mutator.mergeEditorState({
                dragSelection: {
                    startX: drag.startX,
                    startY: drag.startY,
                    currentX: event.pageX - this.containerOffset.left,
                    currentY: event.pageY - this.containerOffset.top,
                    selected: getCollisions(this.props.flowState.nodes, {
                        left,
                        top,
                        right,
                        bottom
                    })
                }
            });
        }
    }

    public onMouseUp(event: React.MouseEvent<HTMLDivElement>): void {
        if (this.props.editorState.dragSelection) {
            this.props.editorState.mutator.mergeEditorState({
                dragSelection: {
                    startX: null,
                    startY: null,
                    currentX: null,
                    currentY: null,
                    selected: this.props.editorState.dragSelection.selected
                }
            });

            if (this.props.editorState.dragSelection.selected) {
                this.Plumber.setDragSelection(this.props.editorState.dragSelection.selected);
            }
        }
    }

    public getDragSelectionBox(): JSX.Element {
        if (this.props.editorState.dragSelection && this.props.editorState.dragSelection.startX) {
            return (
                <div
                    data-spec={dragSelectSpecId}
                    className={styles.dragSelection}
                    style={{ ...getDragStyle(this.props.editorState.dragSelection) }}
                />
            );
        }
        return null;
    }

    public render(): JSX.Element {
        return (
            <>
                {this.getSimulator()}
                {this.getDragNode()}
                {this.getNodeEditor()}
                <div
                    ref={this.onRef}
                    id={this.nodeContainerUUID}
                    className={styles.nodeList}
                    data-spec={nodesContainerSpecId}
                    onMouseDown={this.onMouseDown}
                    onMouseMove={this.onMouseMove}
                    onMouseUp={this.onMouseUp}
                    onDoubleClick={this.onDoubleClick}
                >
                    {this.getStickies()}
                    {this.getDragSelectionBox()}
                    <FlowConsumer>{flowState => this.getNodes(flowState.nodes)}</FlowConsumer>
                </div>
            </>
        );
    }
}
export default React.forwardRef((props: any) => (
    <EditorConsumer>
        {editorState => (
            <FlowConsumer>
                {flowState => <Flow {...props} flowState={flowState} editorState={editorState} />}
            </FlowConsumer>
        )}
    </EditorConsumer>
));
