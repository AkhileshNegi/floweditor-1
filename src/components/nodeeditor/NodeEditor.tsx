// TODO: Remove use of Function
// tslint:disable:ban-types
import { react as bindCallbacks } from 'auto-bind';
import * as React from 'react';
import { EditorConsumer, EditorState } from '~/components/context/editor/EditorContext';
import {
    FlowConsumer,
    FlowState,
    LocalizationUpdates
} from '~/components/context/flow/FlowContext';
import { getDraggedFrom } from '~/components/helpers';
import Modal from '~/components/modal/Modal';
import { Type } from '~/config/typeConfigs';
import { AnyAction } from '~/flowTypes';
import { Asset } from '~/services/AssetService';
import { RenderNode } from '~/store/flowContext';
import { NodeEditorSettings } from '~/store/nodeEditor';

export type UpdateLocalizations = (language: string, changes: LocalizationUpdates) => void;

export interface NodeEditorProps {
    plumberConnectExit: Function;
    plumberRepaintForDuration: Function;

    editorState: EditorState;
    flowState: FlowState;
}

export interface FormProps {
    // our two ways of updating
    updateRouter(renderNode: RenderNode): void;
    updateAction(action: AnyAction): void;

    nodeSettings?: NodeEditorSettings;
    typeConfig?: Type;
    onTypeChange?(config: Type): void;
    onClose?(canceled: boolean): void;

    flowState?: FlowState;
}

export interface LocalizationProps {
    nodeSettings?: NodeEditorSettings;
    typeConfig?: Type;
    onClose?(canceled: boolean): void;

    updateLocalizations: UpdateLocalizations;
    language: Asset;
}

export class NodeEditor extends React.Component<NodeEditorProps> {
    constructor(props: NodeEditorProps) {
        super(props);

        bindCallbacks(this, {
            include: [/^close/, /^update/]
        });
    }

    public close(canceled: boolean): void {
        // Make sure we re-wire the old connection
        if (canceled) {
            const dragPoint = getDraggedFrom(this.props.flowState.nodeEditor.originalNode);
            if (dragPoint) {
                const renderNode = this.props.flowState.nodes[dragPoint.nodeUUID];
                for (const exit of renderNode.node.exits) {
                    if (exit.uuid === dragPoint.exitUUID) {
                        // TODO: should this just be taking literal uuids instead of objects?
                        this.props.plumberConnectExit(renderNode.node, exit);
                        break;
                    }
                }
            }
        }

        this.props.flowState.mutator.closeNodeEditor(canceled);
    }

    public render(): JSX.Element {
        const { flowState, editorState } = this.props;

        if (flowState.nodeEditorOpen) {
            const { typeConfig } = flowState;

            // see if we should use the localization form
            if (this.props.editorState.translating) {
                const { localization: LocalizationForm } = typeConfig;

                if (LocalizationForm) {
                    const localizationProps: LocalizationProps = {
                        updateLocalizations: flowState.mutator.updateLocalizations,
                        nodeSettings: flowState.nodeEditor,
                        typeConfig: flowState.typeConfig,
                        language: editorState.language
                    };

                    return (
                        <Modal width="600px" show={flowState.nodeEditorOpen}>
                            <LocalizationForm {...{ ...localizationProps }} />
                        </Modal>
                    );
                }
            }

            const { form: Form } = typeConfig;
            const formProps: FormProps = {
                updateAction: flowState.mutator.updateAction,
                updateRouter: flowState.mutator.updateRouter,
                nodeSettings: flowState.nodeEditor,
                typeConfig: flowState.typeConfig,
                onTypeChange: flowState.mutator.updateTypeConfig,
                onClose: flowState.mutator.closeNodeEditor,
                flowState
            };

            return (
                <Modal width="600px" show={this.props.flowState.nodeEditorOpen}>
                    <Form {...{ ...formProps }} />
                </Modal>
            );
        }
        return null;
    }
}

export default React.forwardRef((props: any) => (
    <EditorConsumer>
        {editorState => (
            <FlowConsumer>
                {flowState => (
                    <NodeEditor {...props} flowState={flowState} editorState={editorState} />
                )}
            </FlowConsumer>
        )}
    </EditorConsumer>
));
