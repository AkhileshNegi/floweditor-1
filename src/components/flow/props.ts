import { FlowState } from '~/components/context/flow/FlowContext';
import { UpdateLocalizations } from '~/components/nodeeditor/NodeEditor';
import { Type } from '~/config/typeConfigs';
import { AnyAction } from '~/flowTypes';
import { Asset } from '~/services/AssetService';
import { RenderNode } from '~/store/flowContext';
import { NodeEditorSettings } from '~/store/nodeEditor';

export interface ActionFormProps {
    // action details
    nodeSettings: NodeEditorSettings;
    typeConfig: Type;

    // update handlers
    updateAction(action: AnyAction): void;

    // modal notifiers
    onTypeChange(config: Type): void;
    onClose(canceled: boolean): void;
}

export interface RouterFormProps {
    nodeSettings: NodeEditorSettings;
    typeConfig: Type;

    // update handlers
    updateRouter(renderNode: RenderNode): void;

    // modal notifiers
    onTypeChange(config: Type): void;
    onClose(canceled: boolean): void;

    flowState?: FlowState;
}

export interface LocalizationFormProps {
    language: Asset;
    nodeSettings: NodeEditorSettings;
    updateLocalizations(languageCode: string, localizations: any[]): UpdateLocalizations;
    onClose(canceled: boolean): void;
}
