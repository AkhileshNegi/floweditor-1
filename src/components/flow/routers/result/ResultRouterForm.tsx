import { react as bindCallbacks } from 'auto-bind';
import * as React from 'react';
import { FlowConsumer } from '~/components/context/flow/FlowContext';
import Dialog, { ButtonSet } from '~/components/dialog/Dialog';
import { RouterFormProps } from '~/components/flow/props';
import CaseList, { CaseProps } from '~/components/flow/routers/caselist/CaseList';
import OptionalTextInput from '~/components/form/optionaltext/OptionalTextInput';
import TypeList from '~/components/nodeeditor/TypeList';
import { fakePropType } from '~/config/ConfigProvider';
import { Asset } from '~/services/AssetService';
import { Result } from '~/store/flowContext';
import { FormState, StringEntry } from '~/store/nodeEditor';

import { nodeToState, stateToNode } from './helpers';

// TODO: Remove use of Function
// tslint:disable:ban-types
export enum InputToFocus {
    args = 'args',
    min = 'min',
    max = 'max',
    exit = 'exit'
}

export interface ResultRouterFormState extends FormState {
    cases: CaseProps[];
    resultName: StringEntry;
    operand: StringEntry;
    result: Result;
}

export const leadInSpecId = 'lead-in';

class ResultRouterForm extends React.Component<RouterFormProps, ResultRouterFormState> {
    public static contextTypes = {
        endpoints: fakePropType,
        assetService: fakePropType
    };

    constructor(props: RouterFormProps) {
        super(props);

        this.state = nodeToState(this.props.nodeSettings);

        bindCallbacks(this, {
            include: [/^handle/]
        });
    }

    private handleUpdateResultName(value: string): void {
        this.setState({ resultName: { value } });
    }

    private handleResultChanged(selected: Asset[]): void {
        // this.setState({ operand: validate('Operand', value, [validateRequired]) });
    }

    private handleCasesUpdated(cases: CaseProps[]): void {
        this.setState({ cases });
    }

    private handleSave(): void {
        if (this.state.valid) {
            this.props.flowState.mutator.addFlowResult(
                this.state.resultName.value,
                this.props.nodeSettings.originalNode.node.uuid
            );
            this.props.updateRouter(stateToNode(this.props.nodeSettings, this.state));
            this.props.onClose(false);
        }
    }

    private getButtons(): ButtonSet {
        return {
            primary: {
                name: 'Ok',
                onClick: this.handleSave
            },
            secondary: { name: 'Cancel', onClick: () => this.props.onClose(true) }
        };
    }

    public renderEdit(): JSX.Element {
        const typeConfig = this.props.typeConfig;
        return (
            <Dialog
                title={typeConfig.name}
                headerClass={typeConfig.type}
                buttons={this.getButtons()}
            >
                <TypeList
                    __className=""
                    initialType={typeConfig}
                    onChange={this.props.onTypeChange}
                />
                <p>
                    If the result...
                    {Object.keys(this.props.flowState.results)}
                </p>

                <CaseList
                    data-spec="cases"
                    cases={this.state.cases}
                    onCasesUpdated={this.handleCasesUpdated}
                />
                <OptionalTextInput
                    name="Result Name"
                    value={this.state.resultName}
                    onChange={this.handleUpdateResultName}
                    toggleText="Save as.."
                    helpText="By naming the result, you can reference it later using @run.results.whatever_the_name_is"
                />
            </Dialog>
        );
    }

    public render(): JSX.Element {
        return this.renderEdit();
    }
}

export default React.forwardRef((props: any) => (
    <FlowConsumer>
        {flowState => <ResultRouterForm {...props} flowState={flowState} />}
    </FlowConsumer>
));
