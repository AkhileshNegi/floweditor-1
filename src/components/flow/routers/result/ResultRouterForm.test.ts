import { RouterFormProps } from '~/components/flow/props';
import { CaseProps } from '~/components/flow/routers/caselist/CaseList';
import ResultRouterForm from '~/components/flow/routers/result/ResultRouterForm';
import { DEFAULT_OPERAND } from '~/components/nodeeditor/constants';
import { Operators } from '~/config/operatorConfigs';
import { Types } from '~/config/typeConfigs';
import { RouterTypes, SwitchRouter } from '~/flowTypes';
import { composeComponentTestUtils, mock } from '~/testUtils';
import { createRenderNode, getRouterFormProps } from '~/testUtils/assetCreators';
import * as utils from '~/utils';

const { setup } = composeComponentTestUtils<RouterFormProps>(
    ResultRouterForm,
    getRouterFormProps(
        createRenderNode({
            actions: [],
            exits: [],
            ui: { position: { left: 0, top: 0 }, type: Types.split_by_run_result }
        })
    )
);

mock(utils, 'createUUID', utils.seededUUIDs());

jest.mock('~/components/flow/FlowContext');

const exitUUID = utils.createUUID();
const redCaseUUID = utils.createUUID();

describe(ResultRouterForm.name, () => {
    it('should render', () => {
        const { wrapper } = setup();
        expect(wrapper).toMatchSnapshot();
    });

    it('initializes', () => {
        const { wrapper } = setup(true, {
            nodeSettings: {
                $set: {
                    originalNode: createRenderNode({
                        actions: [],
                        exits: [{ destination_node_uuid: null, name: 'Other', uuid: exitUUID }],
                        router: {
                            type: RouterTypes.switch,
                            operand: DEFAULT_OPERAND,
                            cases: [
                                {
                                    uuid: redCaseUUID,
                                    type: Operators.has_any_word,
                                    arguments: ['red'],
                                    exit_uuid: null
                                }
                            ],
                            default_exit_uuid: exitUUID,
                            result_name: 'Color'
                        } as SwitchRouter,
                        ui: {
                            position: { left: 0, top: 0 },
                            type: Types.split_by_run_result
                        }
                    })
                }
            }
        });

        expect(wrapper).toMatchSnapshot();
    });

    describe('updates', () => {
        it('should save changes', () => {
            const { instance, props } = setup(true, {
                $merge: { updateRouter: jest.fn(), onClose: jest.fn() }
            });

            instance.handleUpdateResultName('Favorite Color');
            instance.handleCasesUpdated([
                { kase: { type: Operators.has_any_word, arguments: ['red'] }, exitName: 'Red' },
                { kase: { type: Operators.has_any_word, arguments: ['maroon'] }, exitName: 'Red' },
                { kase: { type: Operators.has_any_word, arguments: ['green'] }, exitName: 'Green' }
            ] as CaseProps[]);

            expect(instance.state).toMatchSnapshot();

            instance.handleSave();
            expect(props.onClose).toHaveBeenCalled();
            expect(props.updateRouter).toHaveBeenCalled();
            expect(props.updateRouter).toMatchCallSnapshot();
        });

        it('should cancel', () => {
            const { instance, props } = setup(true, {
                $merge: { updateRouter: jest.fn(), onClose: jest.fn() }
            });

            instance.getButtons().secondary.onClick();
            expect(props.onClose).toHaveBeenCalled();
            expect(props.updateRouter).not.toHaveBeenCalled();
        });
    });
});
