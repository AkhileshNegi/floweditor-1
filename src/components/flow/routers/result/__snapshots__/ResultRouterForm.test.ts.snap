// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultRouterForm initializes 1`] = `
<Dialog
  buttons={
    Object {
      "primary": Object {
        "name": "Ok",
        "onClick": [Function],
      },
      "secondary": Object {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="split_by_run_result"
  title="Split by Result"
>
  <TypeList
    __className=""
    initialType={
      Object {
        "description": "Split by a result in the flow",
        "form": undefined,
        "localization": [Function],
        "localizeableKeys": Array [
          "exits",
          "cases",
        ],
        "name": "Split by Result",
        "type": "split_by_run_result",
      }
    }
    onChange={[MockFunction]}
  />
  <p>
    If the result...
  </p>
  <Component>
    <Component />
  </Component>
  <CaseList
    cases={
      Array [
        Object {
          "exitName": null,
          "kase": Object {
            "arguments": Array [
              "red",
            ],
            "exit_uuid": null,
            "type": "has_any_word",
            "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
          },
          "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
        },
      ]
    }
    data-spec="cases"
    onCasesUpdated={[Function]}
  />
  <OptionalTextInput
    helpText="By naming the result, you can reference it later using @run.results.whatever_the_name_is"
    name="Result Name"
    onChange={[Function]}
    toggleText="Save as.."
    value={
      Object {
        "value": "Color",
      }
    }
  />
</Dialog>
`;

exports[`ResultRouterForm should render 1`] = `
<Dialog
  buttons={
    Object {
      "primary": Object {
        "name": "Ok",
        "onClick": [Function],
      },
      "secondary": Object {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="split_by_run_result"
  title="Split by Result"
>
  <TypeList
    __className=""
    initialType={
      Object {
        "description": "Split by a result in the flow",
        "form": undefined,
        "localization": [Function],
        "localizeableKeys": Array [
          "exits",
          "cases",
        ],
        "name": "Split by Result",
        "type": "split_by_run_result",
      }
    }
    onChange={[MockFunction]}
  />
  <p>
    If the result...
  </p>
  <Component>
    <Component />
  </Component>
  <CaseList
    cases={Array []}
    data-spec="cases"
    onCasesUpdated={[Function]}
  />
  <OptionalTextInput
    helpText="By naming the result, you can reference it later using @run.results.whatever_the_name_is"
    name="Result Name"
    onChange={[Function]}
    toggleText="Save as.."
    value={
      Object {
        "value": "",
      }
    }
  />
</Dialog>
`;

exports[`ResultRouterForm updates should save changes 1`] = `
Object {
  "cases": Array [
    Object {
      "exitName": "Red",
      "kase": Object {
        "arguments": Array [
          "red",
        ],
        "type": "has_any_word",
      },
    },
    Object {
      "exitName": "Red",
      "kase": Object {
        "arguments": Array [
          "maroon",
        ],
        "type": "has_any_word",
      },
    },
    Object {
      "exitName": "Green",
      "kase": Object {
        "arguments": Array [
          "green",
        ],
        "type": "has_any_word",
      },
    },
  ],
  "operand": Object {
    "value": "@run.input",
  },
  "result": null,
  "resultName": Object {
    "value": "Favorite Color",
  },
  "valid": true,
}
`;

exports[`ResultRouterForm updates should save changes 2`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [],
      "exits": Array [
        Object {
          "destination_node_uuid": null,
          "name": "Red",
          "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
        },
        Object {
          "destination_node_uuid": null,
          "name": "Green",
          "uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
        },
        Object {
          "destination_node_uuid": null,
          "name": "Other",
          "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
        },
      ],
      "router": Object {
        "cases": Array [
          Object {
            "arguments": Array [
              "red",
            ],
            "exit_uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
            "type": "has_any_word",
          },
          Object {
            "arguments": Array [
              "maroon",
            ],
            "exit_uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
            "type": "has_any_word",
          },
          Object {
            "arguments": Array [
              "green",
            ],
            "exit_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
            "type": "has_any_word",
          },
        ],
        "default_exit_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
        "operand": "@run.input",
        "result_name": "Favorite Color",
        "type": "switch",
      },
      "uuid": "48e0a64d-3b3c-4e3e-9d95-7844093edc90",
      "wait": null,
    },
    "ui": Object {
      "position": null,
      "type": "split_by_run_result",
    },
  },
]
`;
